src/metabase/upload.clj

```diff
@@ -483,3 +489,133 @@
                               :num-rows    (count (rest rows))}))]
           (snowplow/track-event! ::snowplow/csv-upload-failed api/*current-user-id* fail-stats))
         (throw e)))))
+
+;;; +--------------------------
+;;; |  append to uploaded table
+;;; +--------------------------
+
+(defn- base-type->upload-type
+  "Returns the most specific upload type for the given base type."
+  [base-type]
+  (condp #(isa? %2 %1) base-type
+    :type/Float                  ::float
+    :type/BigInteger             ::int
+    :type/Integer                ::int
+    :type/Boolean                ::boolean
+    :type/DateTimeWithTZ         ::offset-datetime
+    :type/DateTime               ::datetime
+    :type/Date                   ::date
+    :type/Text                   ::text))
+
+(defn- check-schema
+  "Returns a map with two keys:
+    - `parse-rows`, a function that takes a lazy-seq of rows, and returns a lazy-seq of rows. It parses the
+      string values in each row and removes columns that have the same normalized name as the generated columns.
+    - `table-col-names`, a seq of the column names in the table, in the same order as the columns in the CSV file.
+
+  Throws an exception if:
+    - the CSV file contains duplicate column names
+    - the schema of the CSV file does not match the schema of the table"
+  [table header]
+  ;; Assumes table-cols are unique when normalized
+  (let [table-cols           (t2/select :model/Field :table_id (:id table))
+        table-cols-by-normed (-> (into {}
+                                       (map (fn [col]
+                                              [(normalize-column-name (:name col)) col]))
+                                       table-cols)
+                                 (dissoc "_mb_row_id"))
+        normalized-header+header (into []
+                                       (map (fn [col]
+                                              [(normalize-column-name col) col]))
+                                       header)
+        normed first
+        normalized-header (map normed normalized-header+header)
+        normalized-table-cols (map normed table-cols-by-normed)
```

Comment:
```suggestion
        normalized-header-name (map first normalized-header+header)
        normalized-table-cols-name (keys table-cols-by-normed)
```

why did you alias `normed as first`?

Reply:
Good question, it served as documentation that the first element in both `normalized-header+header` and `table-cols-by-normed` is the "normalized" version. But I think that should be obvious enough given the names, so I've accepted this suggestion